vivisect/analysis/generic/symswitchcase.py

```diff
@@ -0,0 +1,1293 @@
+
+        (csemu, cseffs), asp, fullp = self.getSymbolikParts()
+        cplxIdx = csemu.getSymVariable(smplIdx)
+        return cplxIdx
+
+    #### higher level functionality ####
+
+    def getBoundingCons(self, cplxIdx):
+        return [con for con in self.getConstraints() if contains(con, cplxIdx)[0] ]
+
+    def getNormalizedConstraints(self):
+        '''
+        takes in a complix index symbolik state
+        returns a list of tuples of (SYMT_CON_??, offset)
+        '''
+        retcons = []
+        baseIdx, baseoff = self.getBaseSymIdx()
+
+        for con in self.getBoundingCons(baseIdx): # merge the two
+            cons = con.cons
+
+            if not cons.symtype in (SYMT_CON_GT, SYMT_CON_GE, SYMT_CON_LT, SYMT_CON_LE): 
+                #logger.debug("SKIPPING NON-LIMITING: cons = %s", repr(cons))
+                #return None, None, None
+                continue
+            
+            if cons.kids[1].symtype == SYMT_CONST:
+                symvar = cons.kids[0]
+                symcmp = cons.kids[1]
+            
+            elif cons.kids[0].symtype == SYMT_CONST:
+                symcmp = cons.kids[0]
+                symvar = cons.kids[1]
+            
+            else:
+                # neither side of the constraint is a CONST.  this constraint does 
+                # not set static bounds on idx
+                #logger.debug("SKIPPING (non-discrete): cons = %s", repr(cons))
+                continue
+
+            if not contains(symvar, baseIdx):
+                logger.debug("Constraint not based on Index: %r" % cons)
+                continue
+
+            if cons.symtype == SYMT_CON_LT and symcmp.solve() == 0:
+                #logger.debug("SKIPPING Constraint Checking for Zero: %r" % cons)
+                continue
+
+            conthing, consoff = peelIdxOffset(symvar)
+
+            if conthing != baseIdx:
+                logging.info("FAIL: %r  != %r" % (conthing, baseIdx))
+                #raw_input("FAIL: %r  != %r" % (conthing, baseIdx))
+                continue
+
+            logger.debug("GOOD: %r\n\t%r\n\t%r\t%r + %r" % (cons, symvar, conthing, consoff, symcmp))
+            retcons.append((con, cons.symtype, consoff+symcmp.solve()))
+
+        return retcons
+
+    def getBounds(self):
+        '''
+        Determine the lower index, the upper index, and the offset.
+        lower and upper are inclusive.
+        '''
+        if None not in (self.lower, self.upper, self.baseoff):
+            return self.lower, self.upper, self.baseoff
+
+        (csemu, cseffs), asp, fullp = self.getSymbolikParts()
+
+        lower = 0       # the smallest index used.  most often wants to be 0
+        upper = None    # the largest index used.  max=MAX_CASES
+        count = 0
+        baseoff = None
+       
+        try:
+            while lower is None or (lower == 0 and upper is None) or upper <= lower: # FIXME: this will fail badly when it fails.  make this dependent on the codepathgen
```

Comment:
The fixme feels like a very important note.

Reply:
iinm, that means it will catastrophically fail, which means it simply will get out of the analysis, even though the comment could be interpreted as doing bad things.

Reply:
no, not really.  most of my "FIXME"s are really "TODO"s.  one of my many foibles.